### ISSUE TITLE
Logging clarification requesting

### ISSUE THREAD

**Issue by lukemauldinks** [abdolence/opentelemetry-gcloud-trace-rs#31]
I want both traces and logs to be sent to GCP for collection.  I understand that this crate sends traces to GCP.  Do I need to also setup `tracing-stackdriver` to send logs to GCP or does this crate handle both traces and logs?

**Comment by abdolence** [comment]
Hey, yes - you will need to use `tracing-stackdriver` which works as expected, including that you can see correlation between logs and traces. I decided not to export logs in this crate because `tracing-stackdriver` does its job well and to avoid duplicated logs (GKE/GCE environments), since `tracing-stackdriver` instead of API just uses JSON formatted logging.

This is my setup in one of the projects:

```rust
fn init_console_log(
    tracer: opentelemetry_sdk::trace::Tracer,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let telemetry = tracing_opentelemetry::layer().with_tracer(tracer);

    let subscriber = tracing_subscriber::registry::Registry::default()
        .with(tracing_subscriber::fmt::layer())
        .with(tracing_subscriber::EnvFilter::from_str(
            "gcloud_sdk=debug",
        )?)
        .with(telemetry);

    tracing::subscriber::set_global_default(subscriber)?;

    Ok(())
}

fn init_stackdriver_log(
    gcp_project_id: &str,
    tracer: opentelemetry_sdk::trace::Tracer,
) -> Result<(), BoxedError> {
    use tracing_subscriber::prelude::*;
    let telemetry_layer = tracing_opentelemetry::layer().with_tracer(tracer);

    let stackdriver_layer = tracing_stackdriver::layer().with_cloud_trace(
        tracing_stackdriver::CloudTraceConfiguration {
            project_id: gcp_project_id.to_string(),
        },
    );

    let subscriber = tracing_subscriber::Registry::default()
        .with(telemetry_layer)
        .with(stackdriver_layer)
        .with(tracing_subscriber::EnvFilter::from_str(
            "gcloud_sdk=debug",
        )?);
    tracing::subscriber::set_global_default(subscriber).expect("Could not set up global logger");

    Ok(())
}

async fn init_tracing(app_mode: &GlobalAppMode, gcp_project_id: &str) -> Result<(), BoxedError> {
    let tracer = GcpCloudTraceExporterBuilder::new(gcp_project_id.into())
        .with_trace_config(
            opentelemetry_sdk::trace::Config::default()
                .with_id_generator(opentelemetry_sdk::trace::XrayIdGenerator::default()),
        )
        .install()
        .await?;

    match app_mode {
        GlobalAppMode::Production => init_stackdriver_log(gcp_project_id, tracer),
        GlobalAppMode::Development => init_console_log(tracer),
    }
}
```

**Comment by lukemauldinks** [comment]
do you use the `opentelemetry` feature of the `tracing_stackdriver` crate in your setup?

**Comment by abdolence** [comment]
Yep,
Related deps:

```
tracing = "0.1"
tracing-subscriber = { version = "0.3", features = ["env-filter"] }
tracing-stackdriver = { version = "0.9", features = ["opentelemetry"] }
tracing-futures = { version = "0.2", features = ["futures-03", "tokio"] }
tracing-opentelemetry = "0.22"
opentelemetry = { version = "0.21" }
opentelemetry_sdk = { version = "0.21", features = ["rt-tokio"] }
opentelemetry-gcloud-trace = "0.8"
```

**Comment by lukemauldinks** [comment]
Thank you for the clarification.  I have it working locally now.

**Comment by abdolence** [comment]
I think nowadays `tracing-stackdriver` may work for basic scenarios without using this crate just purely based on log parsing, but I'm not sure if the integration is tight and flexible enough (attributes, export sampling and batching, tracing events, etc). 
Also, it will work only on GCE environments since the parsing happens on the Google side.
This crate implements full-featured opentelemetry exporter that works even without logs.

Hope this helps, and let me know if you have other questions or observations.

**Comment by lukemauldinks** [comment]
@abdolence - the example setup code you posted above was super helpful last year.  I am attempting to upgrade my setup to use opentelemetry-gcloud-trace 0.18.0 and I am getting errors.  Can you please post an update to your setup code and deps?

**Comment by abdolence** [comment]
@lukemauldinks the example nowadays lives in the README here: https://github.com/abdolence/opentelemetry-gcloud-trace-rs?tab=readme-ov-file#integration-with-logs

Let me know if it doesn't work for you.

**Comment by lukemauldinks** [comment]
@abdolence 
Thank you for providing the example but I am still not getting the fields `trace` and `spanId` to show up in my logs.  I will provide you with that code that I am using based on the example.

```
async fn setup_logging(
    gcp_project_id: &str,
    log_mode: LogMode,
    use_auto_log_file: bool,
) -> Result<opentelemetry_sdk::trace::SdkTracerProvider, anyhow::Error> {
    // Setup GCP cloud tracer
    let gcp_trace_exporter = GcpCloudTraceExporterBuilder::for_default_project_id().await?;

    let tracer_provider = gcp_trace_exporter.create_provider().await?;
    let tracer: opentelemetry_sdk::trace::Tracer =
        gcp_trace_exporter.install(&tracer_provider).await?;

    opentelemetry::global::set_tracer_provider(tracer_provider.clone());

    // Create a tracing layer with the configured tracer
    match log_mode {
        LogMode::Gcp => {
            use tracing_subscriber::prelude::*;
            let telemetry_layer = tracing_opentelemetry::layer().with_tracer(tracer);

            let stackdriver_layer = tracing_stackdriver::layer().with_cloud_trace(
                tracing_stackdriver::CloudTraceConfiguration {
                    project_id: gcp_project_id.to_string(),
                },
            );

            let subscriber = tracing_subscriber::Registry::default()
                .with(telemetry_layer)
                .with(stackdriver_layer)
                .with(EnvFilter::from_default_env());

            tracing::subscriber::set_global_default(subscriber)
                .context("setting tracing default failed")?;
        }
        _ => {
         // local logging setup code here
       
    };

    Ok(tracer_provider)
}
```

Cargo.toml
```
tracing-subscriber = { version = "=0.3.19", features = ["fmt", "env-filter"] }
tracing-stackdriver = { version = "=0.10.0", features = ["opentelemetry"] }
tracing-opentelemetry = { version = "=0.29.0" }
opentelemetry = "=0.28.0"
opentelemetry_sdk = { version = "=0.28.0", features = ["rt-tokio"] }
opentelemetry-gcloud-trace = "=0.18.0"
opentelemetry-stdout = { version = "=0.28.0", features = ["trace"] }
```

Example log entry:
```
{
  "insertId": "67b4bead000242f48aa422ad",
  "jsonPayload": {
    "message": "Processing request: path=/httpsApiZenPlannerOnRequest, body_size=1654, content_type=Some(\"application/x-www-form-urlencoded;charset=utf-8\")",
    "target": "kidstrong_webrouter::app",
  },
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
      "location": "us-central1",
      "project_id": "XXX",
      "service_name": "AAA",
      "configuration_name": "BBB",
      "revision_name": "CCC"
    }
  },
  "timestamp": "2025-02-18T17:09:01.148100191Z",
  "severity": "INFO",
  "labels": {
    "managed-by": "github-actions",
    "commit-sha": "AAA",
    "instanceId": "BBB"
  },
  "logName": "projects/XXX-infra/logs/run.googleapis.com%2Fstdout",
  "sourceLocation": {
    "file": "XXX/mod.rs",
    "line": "90"
  },
  "receiveTimestamp": "2025-02-18T17:09:01.156045032Z"
}
```

I am most concerned about the old `tracing-stackdriver` dependency because it targets an old open telemetry version.  Should this code work?

**Comment by lukemauldinks** [comment]
I can confirm via GCP trace explorer that the traces are being written to GCP cloud trace.  However, the `traceId` and `spanId` fields are missing from the logs so the relationship between the logs and traces are broken.

**Comment by abdolence** [comment]
Ah, I see the problem.
Yes, seems `tracing_stackdriver` is not working anymore due to missing upgrade to the latest deps:

![Image](https://github.com/user-attachments/assets/fa3dba57-5f9f-4858-b121-eb84e8ecb7f4)

In my case on GCP console I still see the logs on trace (for some reason), but I'd guess it is fixed by Google and they rely on additional fields (timestamp, etc).

Do you want to open a ticket for them?

**Comment by abdolence** [comment]
The project `tracing_stackdriver` also doesn't look too active recently. I may need to reconsider and introduce log support into this crate.

**Comment by lukemauldinks** [comment]
Yes, I have seen at least one PR for `tracing_stackdriver` to help address the issue but it has not been merged for months.
I made my own fork of the crate and updated it and it is working as expected.  It would be great to have logging support in this crate or a similar one that you can keep updated.

**Comment by abdolence** [comment]
Closing the loop on this one. Released in [0.27.0](https://github.com/abdolence/opentelemetry-gcloud-trace-rs/releases/tag/v0.27.0): this crate no longer depends on `tracing-stackdriver` for logs: it ships its own `tracing_subscriber` layer for Google Cloud Logging, so the tracer and the log layer come from one crate and stay in step with the same OpenTelemetry release.

What changed:

- **Correlation uses the public API.** The layer obtains `trace`/`spanId` through `tracing_opentelemetry::get_otel_context`, which tracing-opentelemetry added for exactly this purpose. It no longer depends on a private span extension, which is what silently broke the ids in the setup discussed above.
- **One home for messages.** Span events are no longer exported to Cloud Trace as annotations by default (`GcpCloudTraceExporterBuilder::with_span_events(true)` restores them). Log lines no longer repeat the enclosing span's name and fields; span attributes live on the span, messages live in Cloud Logging.
- **Two sinks.** JSON lines to stdout for GKE, Cloud Run and GCE with a logging agent (feature `logs`, on by default), or the Cloud Logging API for hosts without an agent (feature `logs-api`). One layer uses exactly one of them, so there are no duplicate entries.

Minimal setup:

```rust
let gcp_trace_exporter = GcpCloudTraceExporterBuilder::for_default_project_id().await?;
let tracer_provider = gcp_trace_exporter.create_provider().await?;
let tracer = gcp_trace_exporter.install(&tracer_provider).await?;

let log_layer = opentelemetry_gcloud_trace::logs::GcpCloudLoggingLayerBuilder::new(project_id).build();

let subscriber = tracing_subscriber::Registry::default()
    .with(tracing_subscriber::EnvFilter::from_default_env())
    .with(tracing_opentelemetry::layer().with_tracer(tracer)) // must come before the log layer
    .with(log_layer);
tracing::subscriber::set_global_default(subscriber)?;
```

The `severity` override, `http_request.*`, `labels.*` and `insert_id` conventions from `tracing-stackdriver` are kept, and `operation.*` was added. Details are in the README section [Integration with logs](https://github.com/abdolence/opentelemetry-gcloud-trace-rs#integration-with-logs), and [`examples/logging.rs`](https://github.com/abdolence/opentelemetry-gcloud-trace-rs/blob/master/examples/logging.rs) is the runnable base setup: console logs in development, JSON on GKE / Cloud Run / GCE, selected by one `APP_MODE` switch. Hosts without a logging agent use [`examples/logging-only-api.rs`](https://github.com/abdolence/opentelemetry-gcloud-trace-rs/blob/master/examples/logging-only-api.rs).
